### DISCUSSION TITLE
Its possible to use Neolink with and back to the official NVR ?

### DISCUSSION THREAD

**Discussion by Lacoste313** [thirtythreeforty/neolink#318] [Q&A]
Hello all, and thanks to the dev for the research and project.

I have 16x B800 camera connected directly to the reolink NVR (H3MB18) POE port (10/100) and my router (pfsense) are directly connected to the NVR LAN port (10/100/100).

Its possible to use Neolink in this situation ? i would get the stream flux to go on my nas with frigate (and same time its possible to nvr)

For the moment, my NVR get a 192.168.6.X IP adresse, but my camera (in NVR UI) are on a another network, like a 172.16.25.Y

I thinks iam forced to use a alternate switch between. camera <> NVR to get the correct ip of my camera.

What do you thinks? 

Thanks you :)

**Comment by QuantumEntangledAndy** [comment]
I believe the correct method is to point neolink towards the NVRs IP address and use the channel setting in the camera config to pick which camera to attach to (starting at channel 0)

Neolink may struggle with 16 simultaneous connections though.

**Reply by Lacoste313** [reply]
Thank you for you reply

This is what i have tried yesterday. with no result. I will need to make more try in evening.

For the moment, Neolink running on docker on my NAS and my .toml file are:
```
[[cameras]]
name = "Camera1" # My camera name set in the NVR UI
username = "USER"
password = "PASS"
address = "192.168.3.10:9000" # The adresse IP of the NVR. I can acces to the WEBUI witth this IP
stream = "both"
```

But for now impossible to access  the RTSP stream with VLC (for fast trying)

> Neolink may struggle with 16 simultaneous connections though.

Yes maybe, i need to try this, i run neolink on a syno RS3614xs+ with 32GB of ram. Maybe neolink can work, need to test.

So, for the moment, only 1 camera B800 are pluged in the NVR for my test

Have a good day

**Reply by QuantumEntangledAndy** [reply]
I'll ping @twistedddx he's our expert on NVRs and the channel parameter was his contribution so perhaps he can chime in if I've got the method wrong.

Do you get any messages in neolink about it making a successful login? That would be the first step to confirm that it is connecting to the cameras.

**Reply by twistedddx** [reply]
I would double check your media port for ther NVR. it is often not 9000
settings -> Network ->Advanced -> Port settings -> Media port

You should also have channel_id in your config file per camera. If not defined each camera will default to channel 0 and you may be trying to connect 16 times for the same camera.
I and others see issues with the NVR supporting more than 4x clear/main streams at once.
The official client appears to run everything as a fluent/sub streams until you select a single camera where it switches that 1 to clear/main. You can make the official client connect to all cameras at once as standard/balanced which is a resolution between main and sub.


You could put all the B800 cameras and the NVR on your own POE switch so that all devices live on your 192.168.3.0/24 network.
Eg do not use the POE switch built in to the NVR.
Neolink could then direct connect to the cameras without needing to go through the NVR.
You would need to test if the cameras are happy to stream to Neolink as well as to the NVR at the same time.
You would probably want DHCP reservations for the cameras and then set the IP/username/password in the NVR for the channels etc.


You should also know the NVR supports RTSP itself. You may not need Neolink.
I am not sure what kind of connections limits may apply to RTSP streaming from the NVR, it might be ~4 main streams.
Hopefully your NVR can be updated to v3.1.0.211. At some point Reolink apparently has added support for RTSP streaming the 4k/h265 stream. Previously only h264 would work.

There is an updated API document linked in the first post here:
https://community.reolink.com/topic/4196/reolink-camera-api-user-guide_v7-update-in-sept-2022

This document notes the updated RTSP URL format that doesn't mention h264 or h265 as it did previously.
main stream
rtsp://(user name):(password)@(ip address):554/Preview_(channelnumber)_main

sub stream url
rtsp://(user name):(password)@(ip address):554/Preview_(channelnumber)_sub

**Reply by Lacoste313** [reply]
@QuantumEntangledAndy Thanks for the ping.


After a lot and many try the last night, i get a valid connection in the log:


```
[2022-11-02T13:21:18Z INFO  neolink::rtsp] Camera1: Connecting to camera at Address: 192.168.1.150:9000 | stderr
[2022-11-02T13:21:18Z INFO  neolink::rtsp] Camera1: Logging in | stderr
[2022-11-02T13:21:18Z INFO  neolink::rtsp] Camera1: Connected and logged in | stderr
[2022-11-02T13:21:18Z INFO  neolink::rtsp] Camera1: Starting video stream Sub Stream (Fluent) | stderr
[2022-11-02T13:21:19Z INFO  neolink::rtsp] Camera1: Starting video stream Main Stream (Clear)
```

So,now yes,its connect, but i dont have the rtsp stream. Maybe because i dont have set the channel_id.
I need more investigate this evening


@twistedddx Wow, that a awesome reply, really thanks man. I see more clearly now.


> I would double check your media port for ther NVR. it is often not 9000
> settings -> Network ->Advanced -> Port settings -> Media port

I will check this this evenin/night.

> You should also have channel_id in your config file per camera. If not defined each camera will default to channel 0 and you may be trying to connect 16 times for the same camera.

Owh, i see, i will set channel_id in the conf file

> I and others see issues with the NVR supporting more than 4x clear/main streams at once.
> The official client appears to run everything as a fluent/sub streams until you select a single camera where it switches that 1 to clear/main. You can make the official client connect to all cameras at once as standard/balanced which is a resolution between main and sub.

This is what iam fear before i post. i need to make some test

> You could put all the B800 cameras and the NVR on your own POE switch so that all devices live on your 192.168.3.0/24 network.
> Eg do not use the POE switch built in to the NVR.
> Neolink could then direct connect to the cameras without needing to go through the NVR.

Yes. i thinks its the better option. But, i need to get a POE switch for testing if 16 mainstream can work proprely directy on Neolink

> You would need to test if the cameras are happy to stream to Neolink as well as to the NVR at the same time.

Yep, and in the case of the cam are not happy, its not a problem for me to just dont use the NVR hardware and just put the real cameraIP on my conf file (.toml)

> You should also know the NVR supports RTSP itself. You may not need Neolink.
> I am not sure what kind of connections limits may apply to RTSP streaming from the NVR, it might be ~4 main streams.

I will try this first before buying a POE switch. But if only ~4 mainstream, its not the solution for me.

> Hopefully your NVR can be updated to v3.1.0.211. At some point Reolink apparently has added support for RTSP streaming the 4k/h265 stream. Previously only h264 would work.
Yes, i have upgraded the NVR and the cam test to the latest version (V3.XX.YY). 

> There is an updated API document linked in the first post here:
> https://community.reolink.com/topic/4196/reolink-camera-api-user-guide_v7-update-in-sept-2022

Thanks for the link, i will see that 


Really thanks for you reply,  now i see the way to go, and what. i need to try :)
Have a good day

**Reply by twistedddx** [reply]
> I am not sure what kind of connections limits may apply to RTSP streaming from the NVR, it might be ~4 main streams.
> 

I just checked this on my RLN8-410 N2MB02 NVR running 3.1.0.211_22102413.
I was able to successfully stream 8x 5MP channels at once via the official NVR's RTSP support, eg no Neolink needed. I did not test with sub streams also, only main streams.

**Reply by Lacoste313** [reply]
> > I am not sure what kind of connections limits may apply to RTSP streaming from the NVR, it might be ~4 main streams.
> 
> I just checked this on my RLN8-410 N2MB02 NVR running 3.1.0.211_22102413. I was able to successfully stream 8x 5MP channels at once via the official NVR's RTSP support, eg no Neolink needed. I did not test with sub streams also, only main streams.

That a good news, maybe de RLN16-410 H3MB18 can work with 16 stream simultanously.

For the moment, i have buyed a second hand POE switch for more test when i have the time for.

But i will try before using POE switch with the official NVR RTSP before (i can use the poe switch for the wifi AP)

Thanks for you try
